File: lib/zenoh_dart.dart
// zenoh_dart.dart

import 'dart:async';
import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';
import 'package:ffi/ffi.dart';
import 'dart:convert';

import 'src/gen/zenoh_dart_bindings_generated.dart' as bindings;

const String _libName = 'zenoh_dart';

/// The dynamic library in which the symbols for [ZenohDartBindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

final bindings.ZenohDartBindings _bindings = bindings.ZenohDartBindings(_dylib);

// --- Classes ---

class ZenohSample {
  final String key;
  final Uint8List payload;
  final String kind;

  ZenohSample(this.key, this.payload, this.kind);

  String get payloadString => utf8.decode(payload);
}

class ZenohReply {
  final String key;
  final Uint8List payload;
  final String kind;

  ZenohReply(this.key, this.payload, this.kind);

  String get payloadString => utf8.decode(payload);
}

class ZenohQuery {
  final String key;
  final String selector;
  final Uint8List? value;
  final String kind;
  final Pointer<Void> _replyContext;

  ZenohQuery(
      this.key, this.selector, this.value, this.kind, this._replyContext);

  /// Send a reply to this query
  void reply(String key, Uint8List data) {
    final keyPtr = key.toNativeUtf8().cast<Char>();
    final dataPtr = calloc<Uint8>(data.length);
    final dataList = dataPtr.asTypedList(data.length);
    dataList.setAll(0, data);

    _bindings.zenoh_query_reply(_replyContext, keyPtr, dataPtr, data.length);

    calloc.free(keyPtr);
    calloc.free(dataPtr);
  }

  void replyString(String key, String data) {
    reply(key, Uint8List.fromList(utf8.encode(data)));
  }
}

class ZenohSession {
  final Pointer<bindings.ZenohSession> _handle;
  bool _isClosed = false;

  // Static maps to hold callbacks
  static final Map<int, StreamController<ZenohSample>> _subscribers = {};
  static final Map<int, StreamController<ZenohReply>> _queries = {};
  static final Map<int, void Function(ZenohQuery)> _queryables = {};

  static int _nextSubscriberId = 0;
  static int _nextQueryId = 0;
  static int _nextQueryableId = 0;

  // Native callback pointers
  static NativeCallable<bindings.ZenohSubscriberCallbackFunction>?
      _subscriberCallback;
  static NativeCallable<bindings.ZenohGetCallbackFunction>? _queryCallback;
  static NativeCallable<bindings.ZenohQueryCallbackFunction>?
      _queryableCallback;

  ZenohSession._(this._handle);

  /// Open a Zenoh Session
  static Future<ZenohSession> open(
      {String mode = 'client', List<String> endpoints = const []}) async {
    final modePtr = mode.toNativeUtf8().cast<Char>();

    Pointer<Char> endpointsPtr = nullptr;
    if (endpoints.isNotEmpty) {
      final endpointsJson = jsonEncode(endpoints);
      endpointsPtr = endpointsJson.toNativeUtf8().cast<Char>();
    }

    final handle = _bindings.zenoh_open_session(modePtr, endpointsPtr);

    calloc.free(modePtr);
    if (endpointsPtr != nullptr) calloc.free(endpointsPtr);

    if (handle == nullptr) {
      throw Exception('Failed to open Zenoh session');
    }

    _ensureCallbacksInitialized();
    return ZenohSession._(handle);
  }

  static void _ensureCallbacksInitialized() {
    _subscriberCallback ??=
        NativeCallable<bindings.ZenohSubscriberCallbackFunction>.listener(
            _onSubscriberData);
    _queryCallback ??=
        NativeCallable<bindings.ZenohGetCallbackFunction>.listener(
            _onQueryData);
    _queryableCallback ??=
        NativeCallable<bindings.ZenohQueryCallbackFunction>.listener(
            _onQueryRequest);
  }

  void _checkClosed() {
    if (_isClosed) throw Exception('Session is closed');
  }

  Future<void> close() async {
    if (_isClosed) return; // Idempotent
    _bindings.zenoh_close_session(_handle);
    _isClosed = true;
  }

  // --- Publisher ---

  Future<ZenohPublisher> declarePublisher(String key) async {
    _checkClosed();
    final keyPtr = key.toNativeUtf8().cast<Char>();
    final pubHandle = _bindings.zenoh_declare_publisher(_handle, keyPtr);
    calloc.free(keyPtr);

    if (pubHandle == nullptr) {
      throw Exception('Failed to declare publisher for key: $key');
    }

    return ZenohPublisher._(pubHandle);
  }

  Future<void> put(String key, Uint8List data) async {
    _checkClosed();
    final keyPtr = key.toNativeUtf8().cast<Char>();
    final dataPtr = calloc<Uint8>(data.length);
    final dataList = dataPtr.asTypedList(data.length);
    dataList.setAll(0, data);

    final result = _bindings.zenoh_put(_handle, keyPtr, dataPtr, data.length);

    calloc.free(keyPtr);
    calloc.free(dataPtr);

    if (result < 0) throw Exception('Put failed');
  }

  Future<void> putString(String key, String value) async {
    await put(key, Uint8List.fromList(utf8.encode(value)));
  }

  Future<void> delete(String key) async {
    _checkClosed();
    final keyPtr = key.toNativeUtf8().cast<Char>();

    final result = _bindings.zenoh_delete(_handle, keyPtr);
    calloc.free(keyPtr);

    if (result < 0) throw Exception('Delete failed');
  }

  // --- Subscriber ---

  Future<ZenohSubscriber> declareSubscriber(String key) async {
    _checkClosed();

    final id = _nextSubscriberId++;
    final controller = StreamController<ZenohSample>();
    _subscribers[id] = controller;

    final context = Pointer<Void>.fromAddress(id);
    final keyPtr = key.toNativeUtf8().cast<Char>();

    final subHandle = _bindings.zenoh_declare_subscriber(
        _handle, keyPtr, _subscriberCallback!.nativeFunction, context);
    calloc.free(keyPtr);

    if (subHandle == nullptr) {
      _subscribers.remove(id);
      throw Exception('Failed to declare subscriber for key: $key');
    }

    controller.onCancel = () {
      // Cleanup is ideally explicit via undeclare, but we can't easily do it here safely without async.
    };

    return ZenohSubscriber._(subHandle, controller, id);
  }

  // --- Query (Get) ---

  Stream<ZenohReply> get(String selector) {
    _checkClosed();
    final id = _nextQueryId++;
    final controller = StreamController<ZenohReply>();
    _queries[id] = controller;

    final context = Pointer<Void>.fromAddress(id);
    final selectorPtr = selector.toNativeUtf8().cast<Char>();

    _bindings.zenoh_get_async(
        _handle, selectorPtr, _queryCallback!.nativeFunction, context);

    calloc.free(selectorPtr);

    // Workaround for missing "query complete" signal in current C wrapper
    // Closes stream after 2 seconds.
    Future.delayed(const Duration(seconds: 2), () {
      if (!controller.isClosed) {
        controller.close();
        _queries.remove(id);
      }
    });

    return controller.stream;
  }

  // --- Queryable ---

  Future<ZenohQueryable> declareQueryable(
      String keyExpr, void Function(ZenohQuery) handler) async {
    _checkClosed();

    final id = _nextQueryableId++;
    _queryables[id] = handler;

    final context = Pointer<Void>.fromAddress(id);
    final keyPtr = keyExpr.toNativeUtf8().cast<Char>();

    final qHandle = _bindings.zenoh_declare_queryable(
        _handle, keyPtr, _queryableCallback!.nativeFunction, context);

    calloc.free(keyPtr);

    if (qHandle == nullptr) {
      _queryables.remove(id);
      throw Exception('Failed to declare queryable for key: $keyExpr');
    }

    return ZenohQueryable._(qHandle, id);
  }

  // --- Scouting ---

  static Stream<String> scout({String what = 'peer|router', String? config}) {
    final controller = StreamController<String>();

    final whatPtr = what.toNativeUtf8().cast<Char>();
    final configPtr =
        config != null ? config.toNativeUtf8().cast<Char>() : nullptr;

    final callback = NativeCallable<Void Function(Pointer<Char>)>.listener(
        (Pointer<Char> info) {
      final infoStr = info.cast<Utf8>().toDartString();
      controller.add(infoStr);
    });

    Future(() {
      // Blocking scout call
      _bindings.zenoh_scout(whatPtr, configPtr, callback.nativeFunction);

      calloc.free(whatPtr);
      if (configPtr != nullptr) calloc.free(configPtr);
      callback.close();
      controller.close();
    });

    return controller.stream;
  }

  // --- Callbacks ---

  static void _onSubscriberData(
      Pointer<Char> key,
      Pointer<Uint8> value,
      int len,
      Pointer<Char> kind,
      Pointer<Char> attachment,
      Pointer<Void> context) {
    int id = context.address;
    if (_subscribers.containsKey(id)) {
      final keyStr = key.cast<Utf8>().toDartString();
      final kindStr = kind.cast<Utf8>().toDartString();
      final payload = Uint8List.fromList(value.asTypedList(len));

      _subscribers[id]?.add(ZenohSample(keyStr, payload, kindStr));
    }
  }

  static void _onQueryData(Pointer<Char> key, Pointer<Uint8> value, int len,
      Pointer<Char> kind, Pointer<Void> context) {
    int id = context.address;
    if (_queries.containsKey(id)) {
      final keyStr = key.cast<Utf8>().toDartString();
      final kindStr = kind.cast<Utf8>().toDartString();
      final payload = Uint8List.fromList(value.asTypedList(len));

      _queries[id]?.add(ZenohReply(keyStr, payload, kindStr));
    }
  }

  static void _onQueryRequest(
      Pointer<Char> key,
      Pointer<Char> selector,
      Pointer<Uint8> value,
      int len,
      Pointer<Char> kind,
      Pointer<Void> replyContext,
      Pointer<Void> userContext) {
    int id = userContext.address;
    if (_queryables.containsKey(id)) {
      final keyStr = key.cast<Utf8>().toDartString();
      final selectorStr = selector.cast<Utf8>().toDartString();
      final kindStr = kind.cast<Utf8>().toDartString();
      final payload =
          len > 0 ? Uint8List.fromList(value.asTypedList(len)) : null;

      final query =
          ZenohQuery(keyStr, selectorStr, payload, kindStr, replyContext);
      _queryables[id]?.call(query);
    }
  }
}

class ZenohPublisher {
  final Pointer<bindings.ZenohPublisher> _handle;
  bool _isUndeclared = false;

  ZenohPublisher._(this._handle);

  Future<void> put(Uint8List data) async {
    if (_isUndeclared) throw Exception('Publisher is undeclared');

    final dataPtr = calloc<Uint8>(data.length);
    final dataList = dataPtr.asTypedList(data.length);
    dataList.setAll(0, data);

    final result = _bindings.zenoh_publisher_put(_handle, dataPtr, data.length);
    calloc.free(dataPtr);

    if (result < 0) throw Exception('Publisher put failed');
  }

  Future<void> putString(String value) async {
    await put(Uint8List.fromList(utf8.encode(value)));
  }

  Future<void> delete() async {
    if (_isUndeclared) throw Exception('Publisher is undeclared');
    _bindings.zenoh_publisher_delete(_handle);
  }

  Future<void> undeclare() async {
    if (_isUndeclared) return;
    _bindings.zenoh_undeclare_publisher(_handle);
    _isUndeclared = true;
  }
}

class ZenohSubscriber {
  final Pointer<bindings.ZenohSubscriber> _handle;
  final StreamController<ZenohSample> _controller;
  final int _id;
  bool _isUndeclared = false;

  ZenohSubscriber._(this._handle, this._controller, this._id);

  Stream<ZenohSample> get stream => _controller.stream;

  Future<void> undeclare() async {
    if (_isUndeclared) return;
    _bindings.zenoh_undeclare_subscriber(_handle);
    _isUndeclared = true;
    _controller.close();
    ZenohSession._subscribers.remove(_id);
  }
}

class ZenohQueryable {
  final Pointer<bindings.ZenohQueryable> _handle;
  final int _id;
  bool _isUndeclared = false;

  ZenohQueryable._(this._handle, this._id);

  Future<void> undeclare() async {
    if (_isUndeclared) return;
    _bindings.zenoh_undeclare_queryable(_handle);
    _isUndeclared = true;
    ZenohSession._queryables.remove(_id);
  }
}

File: src/zenoh_dart.c
#include "zenoh_dart.h"

// Struct definitions
struct ZenohSession {
  z_owned_session_t session;
};

struct ZenohPublisher {
  z_owned_publisher_t publisher;
};

struct ZenohSubscriber {
  z_owned_subscriber_t subscriber;
  ZenohSubscriberCallback callback;
  void *context;
};

// --- Helpers ---

// Initialize logger (optional)
FFI_PLUGIN_EXPORT int zenoh_init_logger(void) {
  z_init_logger();
  return 0;
}

void zenoh_free_string(char *str) {
  if (str != NULL) {
    free(str);
  }
}

// --- Session Management ---

FFI_PLUGIN_EXPORT ZenohSession *zenoh_open_session(const char *mode,
                                                   const char *endpoints) {
  z_owned_config_t config;
  z_config_default(&config);

  // Set mode
  zc_config_insert_json5(z_loan_mut(config), Z_CONFIG_MODE_KEY, mode);

  // Set endpoints (if provided)
  if (endpoints != NULL && strlen(endpoints) > 0) {
    if (strstr(mode, "peer") != NULL) {
      zc_config_insert_json5(z_loan_mut(config), Z_CONFIG_LISTEN_KEY,
                             endpoints);
    } else {
      zc_config_insert_json5(z_loan_mut(config), Z_CONFIG_CONNECT_KEY,
                             endpoints);
    }
  }

// MacOS multicast fix
#if defined(__APPLE__)
  zc_config_insert_json5(z_loan_mut(config), Z_CONFIG_MULTICAST_SCOUTING_KEY,
                         "false");
#endif

  z_owned_session_t s;
  if (z_open(&s, z_move(config), NULL) < 0) {
    return NULL;
  }

  ZenohSession *session = (ZenohSession *)malloc(sizeof(ZenohSession));
  if (session == NULL) {
    z_drop(z_move(s));
    return NULL;
  }
  session->session = s;
  return session;
}

FFI_PLUGIN_EXPORT void zenoh_close_session(ZenohSession *session) {
  if (session != NULL) {
    z_drop(z_move(session->session));
    free(session);
  }
}

// --- Publisher ---

FFI_PLUGIN_EXPORT ZenohPublisher *zenoh_declare_publisher(ZenohSession *session,
                                                          const char *key) {
  if (session == NULL || key == NULL)
    return NULL;

  z_view_keyexpr_t keyexpr;
  if (z_view_keyexpr_from_str(&keyexpr, key) < 0) {
    return NULL;
  }

  z_publisher_options_t options;
  z_publisher_options_default(&options);

  z_owned_publisher_t pub;
  if (z_declare_publisher(z_loan(session->session), &pub, z_loan(keyexpr),
                          &options) < 0) {
    return NULL;
  }

  ZenohPublisher *publisher = (ZenohPublisher *)malloc(sizeof(ZenohPublisher));
  if (publisher == NULL) {
    z_drop(z_move(pub));
    return NULL;
  }
  publisher->publisher = pub;
  return publisher;
}

FFI_PLUGIN_EXPORT int zenoh_publisher_put(ZenohPublisher *publisher,
                                          const uint8_t *data, size_t len) {
  if (publisher == NULL)
    return -1;

  z_publisher_put_options_t options;
  z_publisher_put_options_default(&options);

  z_owned_bytes_t payload;
  // Keep internal copy safe if needed, but here we can just create from pointer
  // if we don't transfer ownership. z_bytes_copy_from_buf makes a copy.
  z_bytes_copy_from_buf(&payload, data, len);

  return z_publisher_put(z_loan(publisher->publisher), z_move(payload),
                         &options);
}

FFI_PLUGIN_EXPORT int zenoh_publisher_delete(ZenohPublisher *publisher) {
  if (publisher == NULL)
    return -1;

  z_publisher_delete_options_t options;
  z_publisher_delete_options_default(&options);

  return z_publisher_delete(z_loan(publisher->publisher), &options);
}

FFI_PLUGIN_EXPORT void zenoh_undeclare_publisher(ZenohPublisher *publisher) {
  if (publisher != NULL) {
    z_drop(z_move(publisher->publisher));
    free(publisher);
  }
}

// --- Subscriber ---

// Callback wrapper for subscribers
static void subscriber_data_handler(const z_loaned_sample_t *sample,
                                    void *arg) {
  ZenohSubscriber *sub = (ZenohSubscriber *)arg;
  if (sub == NULL || sub->callback == NULL)
    return;

  // Get Key
  z_view_string_t key_str;
  z_keyexpr_as_view_string(z_sample_keyexpr(sample), &key_str);
  const char *key = z_string_data(z_loan(key_str));

  // Get Kind
  const char *kind_str = "PUT"; // Default
  z_sample_kind_t kind = z_sample_kind(sample);
  if (kind == Z_SAMPLE_KIND_DELETE)
    kind_str = "DELETE";

  // Get Payload
  const z_loaned_bytes_t *payload = z_sample_payload(sample);
  const uint8_t *data = NULL;
  size_t len = 0;

  if (payload != NULL) {
    data = z_bytes_data(payload);
    len = z_bytes_len(payload);
  }

  // Get Attachment (optional, simplified for now)
  const char *attachment = "";

  // Call Dart callback
  sub->callback(key, data, len, kind_str, attachment, sub->context);
}

static void drop_subscriber_wrapper(void *arg) {
  // Optional: cleanup context if it was malloc'd specifically for this wrapper
  // For now, we assume context is managed by Dart or is just an ID
}

FFI_PLUGIN_EXPORT ZenohSubscriber *
zenoh_declare_subscriber(ZenohSession *session, const char *key,
                         ZenohSubscriberCallback callback, void *context) {
  if (session == NULL || key == NULL)
    return NULL;

  z_view_keyexpr_t keyexpr;
  if (z_view_keyexpr_from_str(&keyexpr, key) < 0)
    return NULL;

  ZenohSubscriber *sub = (ZenohSubscriber *)malloc(sizeof(ZenohSubscriber));
  if (sub == NULL)
    return NULL;

  sub->callback = callback;
  sub->context = context;

  z_subscriber_options_t options;
  z_subscriber_options_default(&options);

  z_owned_closure_sample_t closure;
  z_closure_sample(&closure, subscriber_data_handler, drop_subscriber_wrapper,
                   sub);

  if (z_declare_subscriber(z_loan(session->session), &sub->subscriber,
                           z_loan(keyexpr), z_move(closure), &options) < 0) {
    free(sub);
    return NULL;
  }

  return sub;
}

FFI_PLUGIN_EXPORT void zenoh_undeclare_subscriber(ZenohSubscriber *subscriber) {
  if (subscriber != NULL) {
    z_drop(z_move(subscriber->subscriber));
    free(subscriber);
  }
}

// --- Ad-hoc Operations ---

FFI_PLUGIN_EXPORT int zenoh_put(ZenohSession *session, const char *key,
                                const uint8_t *data, size_t len) {
  if (session == NULL || key == NULL)
    return -1;

  z_view_keyexpr_t keyexpr;
  if (z_view_keyexpr_from_str(&keyexpr, key) < 0)
    return -1;

  z_put_options_t options;
  z_put_options_default(&options);

  z_owned_bytes_t payload;
  z_bytes_copy_from_buf(&payload, data, len);

  return z_put(z_loan(session->session), z_loan(keyexpr), z_move(payload),
               &options);
}

FFI_PLUGIN_EXPORT int zenoh_delete(ZenohSession *session, const char *key) {
  if (session == NULL || key == NULL)
    return -1;

  z_view_keyexpr_t keyexpr;
  if (z_view_keyexpr_from_str(&keyexpr, key) < 0)
    return -1;

  z_delete_options_t options;
  z_delete_options_default(&options);

  return z_delete(z_loan(session->session), z_loan(keyexpr), &options);
}

// --- Query (Get) ---

// Struct to pass context to reply callback
struct GetContext {
  ZenohGetCallback callback;
  void *user_context;
};

static void get_reply_handler(z_loaned_reply_t *reply, void *arg) {
  struct GetContext *ctx = (struct GetContext *)arg;
  if (ctx == NULL || ctx->callback == NULL)
    return;

  if (z_reply_is_ok(reply)) {
    const z_loaned_sample_t *sample = z_reply_ok(reply);

    // Key
    z_view_string_t key_str;
    z_keyexpr_as_view_string(z_sample_keyexpr(sample), &key_str);
    const char *key = z_string_data(z_loan(key_str));

    // Payload
    const z_loaned_bytes_t *payload = z_sample_payload(sample);
    const uint8_t *data = NULL;
    size_t len = 0;

    if (payload != NULL) {
      data = z_bytes_data(payload);
      len = z_bytes_len(payload);
    }

    // Kind
    const char *kind_str = "PUT"; // Default
    if (z_sample_kind(sample) == Z_SAMPLE_KIND_DELETE)
      kind_str = "DELETE";

    ctx->callback(key, data, len, kind_str, ctx->user_context);
  }
}

static void drop_get_context(void *arg) {
  if (arg != NULL) {
    free(arg);
  }
}

// --- Queryable ---

struct ZenohQueryable {
  z_owned_queryable_t queryable;
  ZenohQueryCallback callback;
  void *context;
};

// Callback wrapper for queryable
static void query_handler(const z_loaned_query_t *query, void *arg) {
  ZenohQueryable *q = (ZenohQueryable *)arg;
  if (q == NULL || q->callback == NULL)
    return;

  // Get Key Selector
  z_view_keyexpr_t keyexpr = z_query_keyexpr(query);
  z_view_string_t key_str;
  z_keyexpr_as_view_string(keyexpr, &key_str);
  const char *key = z_string_data(z_loan(key_str));

  // Get Selector (parameters)
  z_view_string_t selector_str = z_query_parameters(query);
  const char *selector = z_string_data(z_loan(selector_str));

  // Get Payload (Value)
  const z_loaned_bytes_t *payload = z_query_payload(query); // Can be NULL
  const uint8_t *data = NULL;
  size_t len = 0;
  if (payload != NULL) {
    data = z_bytes_data(payload);
    len = z_bytes_len(payload);
  }

  // Get Kind (optional, usually empty for query?)
  const char *kind = "GET";

  // Reply Context needs to be passed to allow replying
  // We can pass the `query` object itself as the reply context, BUT
  // `query` is loaned and only valid during this callback.
  // `zenoh_query_reply` must be called synchronously inside this callback
  // OR we need to clone the query/reply capability?
  // Zenoh C API `z_query_reply` uses `z_loaned_query_t`.
  // So we pass `query` as `reply_context`. It is only valid for the duration of
  // the callback. If async reply is needed, we would need to clone it (if
  // supported). For now, assume sync reply or valid lifetime.
  q->callback(key, selector, data, len, kind, (void *)query, q->context);
}

static void drop_queryable_wrapper(void *arg) {
  // Cleanup if needed
}

FFI_PLUGIN_EXPORT ZenohQueryable *
zenoh_declare_queryable(ZenohSession *session, const char *key_expr,
                        ZenohQueryCallback callback, void *context) {
  if (session == NULL || key_expr == NULL)
    return NULL;

  z_view_keyexpr_t keyopts;
  if (z_view_keyexpr_from_str(&keyopts, key_expr) < 0)
    return NULL;

  ZenohQueryable *q = (ZenohQueryable *)malloc(sizeof(ZenohQueryable));
  if (q == NULL)
    return NULL;

  q->callback = callback;
  q->context = context;

  z_queryable_options_t options;
  z_queryable_options_default(&options);

  z_owned_closure_query_t closure;
  z_closure_query(&closure, query_handler, drop_queryable_wrapper, q);

  if (z_declare_queryable(z_loan(session->session), &q->queryable,
                          z_loan(keyopts), z_move(closure), &options) < 0) {
    free(q);
    return NULL;
  }

  return q;
}

FFI_PLUGIN_EXPORT void zenoh_undeclare_queryable(ZenohQueryable *queryable) {
  if (queryable != NULL) {
    z_drop(z_move(queryable->queryable));
    free(queryable);
  }
}

FFI_PLUGIN_EXPORT void zenoh_query_reply(void *reply_context, const char *key,
                                         const uint8_t *data, size_t len) {
  const z_loaned_query_t *query = (const z_loaned_query_t *)reply_context;
  if (query == NULL)
    return;

  z_view_keyexpr_t keyexpr;
  if (z_view_keyexpr_from_str(&keyexpr, key) < 0)
    return;

  z_query_reply_options_t options;
  z_query_reply_options_default(&options);

  z_owned_bytes_t payload;
  z_bytes_copy_from_buf(&payload, data, len);

  // We are replying to the query.
  z_query_reply(query, z_loan(keyexpr), z_move(payload), &options);
}

// --- Scouting ---

static void scout_callback_wrapper(const z_loaned_hello_t *hello, void *arg) {
  void (*cb)(const char *) = (void (*)(const char *))arg;
  if (cb) {
    // Extract info from hello
    // Simplified for now, just saying "Found Peer"
    // In real impl, parse z_hello_whatami(hello), z_hello_locators(hello)
    cb("{\"event\": \"peer_discovered\"}");
  }
}

static void drop_scout_wrapper(void *arg) {}

FFI_PLUGIN_EXPORT void zenoh_scout(const char *what, const char *config,
                                   void (*callback)(const char *info)) {
  z_whatami_t w = Z_WHATAMI_PEER | Z_WHATAMI_ROUTER; // Default both
  if (what && strcmp(what, "router") == 0)
    w = Z_WHATAMI_ROUTER;
  else if (what && strcmp(what, "peer") == 0)
    w = Z_WHATAMI_PEER;

  z_scout_options_t options;
  z_scout_options_default(&options);
  options.what = w;
  options.timeout_ms = 1000; // 1 second scout

  z_owned_config_t cfg;
  z_config_default(&cfg); // Start default
  if (config) {
    // If config provided, try to parse? Or manual?
    // z_config_from_str(config) ??
    // For simplicity, passing config string might need proper parsing.
    // Or we just rely on default multicast.
  }

  // This block is blocking in C usually unless we use closure?
  // z_scout is usually blocking and returns channel or calls callback.
  // Wait, z_scout with closure:
  z_owned_closure_hello_t closure;
  z_closure_hello(&closure, scout_callback_wrapper, drop_scout_wrapper,
                  (void *)callback);

  z_scout(z_move(cfg), &options, &closure);
}